packages/application/src/shell.ts

```diff
@@ -1629,13 +1629,14 @@ export class LabShell extends Widget implements JupyterFrontEnd.IShell {
     args: FocusTracker.IChangedArgs<Widget>
   ): void {
     if (args.newValue) {
-      args.newValue.title.className += ` ${CURRENT_CLASS}`;
+      args.newValue.title.className = `${args.newValue.title.className} ${CURRENT_CLASS} ${ACTIVE_CLASS}`;
+      args.newValue.activate();
+      this._activeChanged.emit(args);
     }
     if (args.oldValue) {
-      args.oldValue.title.className = args.oldValue.title.className.replace(
-        CURRENT_CLASS,
-        ''
-      );
+      args.oldValue.title.className = args.oldValue.title.className
+        .replace(CURRENT_CLASS, '')
+        .replace(ACTIVE_CLASS, '');
```

Comment:
Could we use [`classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) instead? The current implementation will lead to a slow memory leak (two spaces at a time after each activation/deactivation of a tab, and then each replace will take longer because it naively scans each character; I don't imaging it to be a big problem by itself by these tiny things can compound).

Reply:
@krassowski Can't access `classList` methods from `title`. They are available on parent widgets `newValue` and `oldValue` but using them as shown in code below doesn't actually target the appropriate DOM element and doesn't work.

```ts
  /**
   * Handle a change to the dock area current widget.
   */
  private _onCurrentChanged(
    sender: any,
    args: FocusTracker.IChangedArgs<Widget>
  ): void {
    if (args.newValue) {
      args.newValue.addClass(CURRENT_CLASS);
      args.newValue.addClass(ACTIVE_CLASS);
      args.newValue.activate();
      this._activeChanged.emit(args);
    }
    if (args.oldValue) {
      args.oldValue.removeClass(CURRENT_CLASS);
      args.oldValue.removeClass(ACTIVE_CLASS);
    }
    this._currentChanged.emit(args);
    this._onLayoutModified();
  }
```

Reply:
In that case I guess we could go for simpler:

```suggestion
      args.oldValue.title.className = args.oldValue.title.className
        .replace(CURRENT_CLASS, '')
        .replace(ACTIVE_CLASS, '')
        .replace('  ', ' ');
```

but of course we could also add `classList` to [`Title`](https://lumino.readthedocs.io/en/latest/api/classes/widgets.Title-1.html#className)

Reply:
@krassowski thanks for pointing this our. Came up how to resolve this without additional replace by always clearing preceding space together with class name (`.replace(` ${CURRENT_CLASS}`, '')`). Also changed test to make sure classes are removed cleanly by checking that className string after widget was activated and deactivated is equal to original className string before the widget was touched

Reply:
I also created lumino [#539 ](https://github.com/jupyterlab/lumino/issues/539) to track adding classList to` widget.title` as having it would remove the necessity for raw string manipulation that is prone to edge cases and inefficiencies